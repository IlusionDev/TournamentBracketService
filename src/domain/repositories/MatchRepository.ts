import { getRepository, Repository } from "typeorm";
import Match from "@/domain/models/Match";


export default class MatchRepository {
    static getRepository() {
        return getRepository(Match);
    }
}
