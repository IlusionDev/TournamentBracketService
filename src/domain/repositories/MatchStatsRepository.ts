import { getRepository, Repository } from "typeorm";
import MatchStats  from "@/domain/models/MatchStats";


export default class MatchStatsRepository {


    static getRepository() {
        return getRepository(MatchStats);
    }

    async getAllMatchStatsByMatch(matchId: string, playersId: string[]) {
        return await getRepository(MatchStats).createQueryBuilder("matchStats")
            .where("matchStats.match_id = :matchId", { matchId })
            .andWhere("matchStats.player_id IN (:...playersId)", { playersId })
            .getMany();
    }
}
