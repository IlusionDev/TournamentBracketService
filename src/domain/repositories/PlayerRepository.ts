import { getRepository, Repository } from "typeorm";
import Player from "@/domain/models/Player";


export default class PlayerRepository {


    static getRepository() {
        return getRepository(Player);
    }

    findByEmail(email: string) {
        return getRepository(Player).findOne({ where: { email } });
    }

}
