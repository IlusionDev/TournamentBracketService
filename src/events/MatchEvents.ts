import { ProjectEvent } from "@/events/ProjectEvent";
import MatchStats from "@/domain/models/MatchStats";

export default class MatchEvents extends ProjectEvent {

    private static instance: MatchEvents;
    public static readonly finishedEvent: string = "matchFinished";
    public static readonly statsUpdatedEvent: string = "matchStatsUpdated";
    public static readonly createTiedBreakerEvent: string = "createTiedBreaker";
    private static events = {
        [MatchEvents.finishedEvent]: MatchEvents.finishedEvent,
        [MatchEvents.statsUpdatedEvent]: MatchEvents.statsUpdatedEvent,
        [MatchEvents.createTiedBreakerEvent]: MatchEvents.createTiedBreakerEvent
    };

    constructor() {
        super(MatchEvents.events);
    }

    static getInstance() {
        if (!MatchEvents.instance) {
            MatchEvents.instance = new MatchEvents();
        }

        return MatchEvents.instance;
    }

    public finished(roundId: any): void {
        this.emit(MatchEvents.finishedEvent, roundId);
    }

    public statsUpdated(roundId: any): void {
        this.emit(MatchEvents.statsUpdatedEvent, roundId);
    }

    public createTieBreaker(roundId: string, playerIds: string[]): void {
        this.emit(MatchEvents.createTiedBreakerEvent, roundId, playerIds);
    }
}

