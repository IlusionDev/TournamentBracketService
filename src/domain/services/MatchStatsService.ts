import MatchStats from "@/domain/models/MatchStats";
import Match from "@/domain/models/Match";
import Player from "@/domain/models/Player";
import { ErrorNotFound } from "@/global/errors/ErrorNotFound";
import { ErrorNotUpdated } from "@/global/errors/ErrorNotUpdated";
import { ErrorNotSaved } from "@/global/errors/ErrorNotSaved";
import MatchEvents from "@/events/MatchEvents";
import { SaveMatchStatsDto } from "@/domain/definitions/dto/SaveMatchStatsDto";
import { UpdateMatchStatsDto } from "@/domain/definitions/dto/UpdateMatchStatsDto";
import MatchRepository from "@/domain/repositories/MatchRepository";
import PlayerRepository from "@/domain/repositories/PlayerRepository";
import MatchStatsRepository from "@/domain/repositories/MatchStatsRepository";

export class MatchStatsService {

    async _haveMatchStatsInMatch(player: Player, match: Match) {
        const matchStatsRepository = MatchStatsRepository.getRepository();

        const matchStats = await matchStatsRepository.findOne({
            where: {
                player,
                match
            }
        });

        return matchStats !== undefined;
    }

    async _createMatchStatsConditions(player: Player, match: Match) {
        if (!match) {
            throw new ErrorNotFound('Match');
        }

        if (!player) {
            throw new ErrorNotFound('Player');
        }

        const haveMatchStatsInBracket = await this._haveMatchStatsInMatch(player, match);

        if (haveMatchStatsInBracket) {
            throw new ErrorNotSaved("Match stats already exists");
        }
    }

    async saveMatchStats(saveMatchStatsDto: SaveMatchStatsDto) {
        const matchRepository = MatchRepository.getRepository();
        const playerRepository = PlayerRepository.getRepository();
        const matchStatsRepository = MatchStatsRepository.getRepository();

        const match: Match = await matchRepository.findOne(saveMatchStatsDto.matchId);
        const player: Player = await playerRepository.findOne(saveMatchStatsDto.playerId);

        await this._createMatchStatsConditions(player, match);

        const newMatchStats = matchStatsRepository.create({
            match,
            points: saveMatchStatsDto.points,
            player
        } as any);

        const savedMatchStats = await matchStatsRepository.save(newMatchStats);

        if (!savedMatchStats) {
            throw new ErrorNotSaved('MatchStats');
        }

        return savedMatchStats
    }

    async updateMatchStats(updateMatchStatsDto: UpdateMatchStatsDto) {
        const matchStatsRepository = MatchStatsRepository.getRepository();
        const matchStats: MatchStats = await matchStatsRepository.findOne(updateMatchStatsDto.id);

        if (!matchStats) {
            throw new ErrorNotFound('MatchStats');
        }
        const match = await matchStats.match
        const round = await match.round

        if (round.finished) {
            throw new ErrorNotUpdated('Match round already finished')
        }

        if (!!updateMatchStatsDto.finished) {
            matchStats.finished = true;
        } else if (updateMatchStatsDto.points) {
            matchStats.points += updateMatchStatsDto.points;
        }

        const updatedMatchStats = await matchStatsRepository.save(matchStats);

        if (!updatedMatchStats) {
            throw new ErrorNotUpdated('MatchStats');
        }

        MatchEvents.getInstance().statsUpdated(match.id);

        return updatedMatchStats;
    }

    async getMatchStats(matchStatsId: number) {
        const matchStatsRepository = MatchStatsRepository.getRepository();
        const matchStats: MatchStats = await matchStatsRepository.findOne(matchStatsId);

        if (!matchStats) {
            throw new ErrorNotFound('MatchStats');
        }

        return matchStats;
    }

    async getAllMatchStatusByPlayer(playerId: number) {
        const matchStatsRepository = MatchStatsRepository.getRepository();
        const matchStats: MatchStats[] = await matchStatsRepository.find({
            where: {
                player: {
                    id: playerId
                }
            }
        });

        if (!matchStats) {
            throw new ErrorNotFound('MatchStats');
        }

        return matchStats;
    }

}
