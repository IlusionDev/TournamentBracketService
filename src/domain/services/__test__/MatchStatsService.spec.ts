import MatchRepository from "../../repositories/MatchRepository";
import PlayerRepository from "../../repositories/PlayerRepository";
import {MatchStatsService} from "../MatchStatsService";
import MatchStatsRepository from "../../repositories/MatchStatsRepository";
import MatchEvents from "../../../events/MatchEvents";

const round = {
    finished: false,
}

const player = {
    id: "playerId",
}

const match = {
    id: "matchId",
    round: round,
    players: [player],
}

const matchStatsMockRepository = {
    findOne: jest.fn(() => new Promise((resolve) => resolve({
        match,
    }))),
    create: jest.fn((data) => data),
    save: jest.fn((data) => new Promise((resolve) => resolve(data))),
}

const matchMockRepository = {
    create: jest.fn((data) => data),
    save: jest.fn((data) => new Promise((resolve) => resolve(data))),
    findOne: jest.fn(() => new Promise((resolve) => resolve({})))
}

const playerMockRepository = {
    findOne: jest.fn(() => new Promise((resolve) => resolve(player))),
    findByIds: jest.fn((data) => new Promise((resolve) => resolve([
        {},
        {},
    ])))
}

const matchEventsMock = {
    statsUpdated: jest.fn()
}

// @ts-ignore
const matchRepository = MatchRepository.getRepository = jest.fn(() => matchMockRepository);
// @ts-ignore
const playerRepository = PlayerRepository.getRepository = jest.fn(() => playerMockRepository);
// @ts-ignore
const matchStatsRepository = MatchStatsRepository.getRepository = jest.fn(() => matchStatsMockRepository);
// @ts-ignore
const matchEvents = MatchEvents.getInstance = jest.fn(() => matchEventsMock);

const matchStatsService = new MatchStatsService()

describe("MatchStatsService", () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    it("check if match have stats", async () => {
        const haveStats = await matchStatsService._haveMatchStatsInMatch(player as any, match as any)

        expect(matchStatsMockRepository.findOne).toHaveBeenNthCalledWith(1, {
            where: {
                player,
                match
            }
        })
        expect(haveStats).toEqual(true)
    });

    it("saves the first stat of a player in a match", async () => {
        const saveMatchStatsDto = {
            matchId: "matchId",
            playerId: "playerId",
            points: 2
        }
        const expectedSavedModel = {
            match: {},
            player,
            points: 2
        }
        // @ts-ignore
        matchStatsService._createMatchStatsConditions = jest.fn(() => true)
        const savedModel = await matchStatsService.saveMatchStats(saveMatchStatsDto)

        expect(matchMockRepository.findOne).toHaveBeenNthCalledWith(1, "matchId")
        expect(playerMockRepository.findOne).toHaveBeenNthCalledWith(1, "playerId")
        expect(matchStatsMockRepository.create).toHaveBeenNthCalledWith(1, expectedSavedModel)
        expect(matchStatsMockRepository.save).toHaveBeenNthCalledWith(1, expectedSavedModel)
        expect(savedModel).toMatchSnapshot()
    });

    it("updates stat of a player in a match", async () => {
        const saveMatchStatsDto = {
            id: "matchStatsId",
            points: 23,
            finished: true
        }
        const expectedSavedModel = {
            match: {},
            player,
            points: 2
        }

        const updatedModel = await matchStatsService.updateMatchStats(saveMatchStatsDto)

        expect(matchStatsMockRepository.findOne).toHaveBeenNthCalledWith(1, "matchStatsId")
        expect(matchStatsMockRepository.save).toHaveBeenNthCalledWith(1, {
            match,
            finished: true
        })
        expect(matchEventsMock.statsUpdated).toHaveBeenNthCalledWith(1, "matchId")
        expect(updatedModel).toMatchSnapshot()
    });
});
