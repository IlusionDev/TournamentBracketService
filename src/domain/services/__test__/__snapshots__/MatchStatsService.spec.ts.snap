// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MatchStatsService saves the first stat of a player in a match 1`] = `
Object {
  "match": Object {},
  "player": Object {
    "id": "playerId",
  },
  "points": 2,
}
`;

exports[`MatchStatsService updates stat of a player in a match 1`] = `
Object {
  "finished": true,
  "match": Object {
    "id": "matchId",
    "players": Array [
      Object {
        "id": "playerId",
      },
    ],
    "round": Object {
      "finished": false,
    },
  },
}
`;
